libs/cuda_support/include/hpx/cuda_support/cuda_future.hpp

```diff
@@ -0,0 +1,343 @@
+        template <typename Allocator>
+        struct future_data
+          : lcos::detail::future_data_allocator<void, Allocator>
+        {
+            HPX_NON_COPYABLE(future_data);
+
+            using init_no_addref =
+                typename lcos::detail::future_data_allocator<void,
+                    Allocator>::init_no_addref;
+
+            using other_allocator = typename std::allocator_traits<
+                Allocator>::template rebind_alloc<future_data>;
+
+            future_data()
+              : rt_(hpx::get_runtime_ptr())
+            {
+            }
+
+            // constructor used by callback based futures
+            future_data(init_no_addref no_addref, other_allocator const& alloc,
+                cudaStream_t stream, callback_mode)
+              : lcos::detail::future_data_allocator<void, Allocator>(
+                    no_addref, alloc)
+              , rt_(hpx::get_runtime_ptr())
+            {
+                init_callback(stream);
+            }
+
+            // constructor used by event based futures
+            future_data(init_no_addref no_addref, other_allocator const& alloc,
+                cudaStream_t stream, event_mode)
+              : lcos::detail::future_data_allocator<void, Allocator>(
+                    no_addref, alloc)
+            {
+                init_event(stream);
+            }
+
+            // @TODO - another thread might queue a kernel function
+            // at the same time, on the same stream - causing our event
+            // to be later than expected :
+            // should there be a mutex protecting future creation + event record?
+            void init_event(cudaStream_t stream)
+            {
+                // even if the event is triggered/ready before we complete, we do
+                // not need to worry about a race, because we only poll for the complete
+                // event after returning from here
+                if (!cuda_event_pool::get_event_pool().pop(event_))
+                {
+                    HPX_THROW_EXCEPTION(invalid_status, "init_event",
+                        "cuda event stack size has been exceeded "
+                        "recompile with a larger event stack size");
+                }
+                check_cuda_error(cudaEventRecord(event_, stream));
+
+                cud_debug.debug(
+                    debug::str<>("init_event"), "event", debug::hex<8>(event_));
+            }
+
+            void init_callback(cudaStream_t stream)
+            {
+                // Hold on to the shared state on behalf of the cuda runtime
+                // right away as the callback could be called immediately
+                // and it happens on another thread which might cause a race
+                lcos::detail::intrusive_ptr_add_ref(this);
+
+                cudaError_t error =
+                    cudaStreamAddCallback(stream, stream_callback, this, 0);
+                if (error != cudaSuccess)
```

Comment:
Since `cudaStreamAddCallback()` is deprecated, have you considered using `cudaLaunchHostFunc` instead?

Reply:
Very good idea. I will try that.

Reply:
Marking this as resolved as I will look into it in the next PR ...